monitoring/grafana/WDSAPIServerMonitoring-1722629699305.json

```diff
@@ -0,0 +1,686 @@
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "prometheus-test"
+          },
+          "disableTextWrap": false,
+          "editorMode": "code",
+          "expr": "irate(process_cpu_seconds_total{job=~\"$SpaceName\", namespace=\"$SpaceNS\"}[1m])",
+          "fullMetaSearch": false,
+          "includeNullMetadata": true,
+          "instant": false,
+          "legendFormat": "__auto",
+          "range": true,
+          "refId": "A",
+          "useBackend": false
+        }
+      ],
+      "title": "process_cpu_seconds_total",
+      "type": "timeseries"
+    },
+    {
+      "datasource": {
+        "default": true,
+        "type": "prometheus",
+        "uid": "prometheus-test"
+      },
+      "fieldConfig": {
+        "defaults": {
+          "color": {
+            "mode": "palette-classic"
+          },
+          "custom": {
+            "axisBorderShow": false,
+            "axisCenteredZero": false,
+            "axisColorMode": "text",
+            "axisLabel": "",
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "barWidthFactor": 0.6,
+            "drawStyle": "line",
+            "fillOpacity": 0,
+            "gradientMode": "none",
+            "hideFrom": {
+              "legend": false,
+              "tooltip": false,
+              "viz": false
+            },
+            "insertNulls": false,
+            "lineInterpolation": "linear",
+            "lineWidth": 1,
+            "pointSize": 5,
+            "scaleDistribution": {
+              "type": "linear"
+            },
+            "showPoints": "auto",
+            "spanNulls": false,
+            "stacking": {
+              "group": "A",
+              "mode": "none"
+            },
+            "thresholdsStyle": {
+              "mode": "off"
+            }
+          },
+          "mappings": [],
+          "thresholds": {
+            "mode": "absolute",
+            "steps": [
+              {
+                "color": "green",
+                "value": null
+              },
+              {
+                "color": "red",
+                "value": 80
+              }
+            ]
+          }
+        },
+        "overrides": []
+      },
+      "gridPos": {
+        "h": 8,
+        "w": 12,
+        "x": 12,
+        "y": 8
+      },
+      "id": 4,
+      "options": {
+        "legend": {
+          "calcs": [],
+          "displayMode": "list",
+          "placement": "bottom",
+          "showLegend": true
+        },
+        "tooltip": {
+          "mode": "single",
+          "sort": "none"
+        }
+      },
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "prometheus-test"
+          },
+          "disableTextWrap": false,
+          "editorMode": "builder",
+          "expr": "sum by(le) (increase(etcd_request_duration_seconds_bucket{job=\"$SpaceName\", namespace=\"$SpaceNS\"}[1m]))",
```

Comment:
So this query is producing the amount by which each bucket's count increased over the last minute. That is an odd thing to present to a user.

Reply:
This query measures HTTP requests duration as measured over the last 1 minutes.  It calculates the increase in the time series in the range vector of 1 minute.

Reply:
`etcd_request_duration_seconds_bucket`, for a particular set of labels, is the count of etcd requests that the apiserver made and got a latency less than or equal to the value of the `le` label.

Try a query such as a bare `etcd_request_duration_seconds_bucket`, with just the basic label restriction and no summation, rate, or histogram function (e.g., `etcd_request_duration_seconds_bucket{namespace="wds1-system", job="wds1"}`). You will get a tremendous number of timeseries. Here is one that I got:

```
etcd_request_duration_seconds_bucket{container="kube-apiserver", endpoint="https", instance="10.244.0.18:9444", job="wds1", le="0.005", namespace="wds1-system", operation="create", pod="kube-apiserver-69fc88ffdc-8pjt8", service="wds1", type="apiservices.apiregistration.k8s.io"}
```

Its value was 17. If you look, you will find that the value in every bucket is always an integer. That is because it is a count of a number of observations.

The query in this panel, `sum by(le) (increase(etcd_request_duration_seconds_bucket{job="$SpaceName", namespace="$SpaceNS"}[1m]))`, measures the change in the count sum, for each bucket boundary, over the course of the last minute (https://prometheus.io/docs/prometheus/latest/querying/functions/#increase).

What you would normally do with a histogram metric is compute the average or a percentile. Have a look at the average, `(rate(etcd_request_duration_seconds_sum[1m])) / (rate(etcd_request_duration_seconds_count[1m]))`.  The following screen capture shows what I got when I did.

<img width="925" alt="Screenshot 2024-08-09 at 2 36 11 AM" src="https://github.com/user-attachments/assets/2d52bbaa-ba7a-4b78-90ba-a0ddeb0fe60d">

Almost all of the averages are below 5 milliseconds. In fact, if you look at those sums of increases, you will find that they are the same for almost all buckets, and the lowest bucket counts latencies `<=` 5ms. In other words, very nearly all of the latencies are under 5ms so the quantiles derived from these histograms are meaningless.

Reply:
Fixed! Update with the query that you proposed above using the `average`.